### ISSUE TITLE
Check if special permissions for js lib are required

### ISSUE THREAD

**Issue by hrj** [UprootLabs/gngr#29]
We had to add the debugger to the window context factory in order to workaround a race condition in the JS library.

But it is commented out in the ugly patches because it seemed to be not required anymore.

However, the race condition appeared again today.

Quick testing shows that using a dummy debugger works fine. But needs more testing.

### TODO
 1. Try to make a reproducible test-case for the race condition
 2. Check if a dummy debugger can be used. If so, remove the extra permission from the ugly patch set.

**Comment by hrj** [comment]
Tests seem to be working fine without the debugger. Removed the extra permission from ugly patches branch.

**Comment by hrj** [comment]
The problem surfaced again on Reddit :disappointed: 

The good news is that the problem seemed to go away after using a dummy debugger that returns dummy frames (instead of just null frames). If this is confirmed, no extra permissions would be required.

Have opened an [upstream thread](https://github.com/apigee/rhino/issues/23)

**Comment by hrj** [comment]
I was able to pin down the root-cause. There is code to set the parent scope of an object, when that object is being returned to JS. Sometimes the `window` object is returned and its parent scope is set to a non-null scope. However, this causes a circular chain of parent-scopes as some of those scopes might themselves have a parent-scope set to `window`.

I was able to determine atleast one instance where this happens, and coded it to [a test-case](http://uprootlabs.github.io/gngrTestCases/test-case/i29/)

I am yet to code the part that causes the infinite loop (if there is no fix to the first part). Anything that walks the parent scope chain should work.

A fix has been committed to the ugly-patches branch. It consists of a simple check before setting the parent-scope of `window`.
